Qt patches

transmitterdan opened the pull request.

bdbcat committed:
Fill in the binaries SHA-1 checksums

bdbcat committed:
Implement wxGLCanvas attribute setting logic for wxQt

bdbcat committed:
Implement wxGLCanvas attribute setting logic for wxQt

bdbcat committed:
Protect against orphan Qt events and signals on wxWindow and wxGLCanvas DTOR.

bdbcat committed:
Protect against orphan Qt events and signals on wxWindow and wxGLCanvas DTOR.

bdbcat committed:
Correct for added localeconv() found in NDK21d

bdbcat committed:
Correct for added localeconv() found in NDK21d

bdbcat committed:
Implement wxMenuItem::SetFont()

bdbcat committed:
Implement wxMenuItem::SetFont()

bdbcat committed:
Correct android_wcsto* for faulty strto* return values.

bdbcat committed:
Correct android_wcsto* for faulty strto* return values.

bdbcat committed:
Enable wxGestureEvent support for Qt

bdbcat committed:
Enable wxGestureEvent support for Qt

vadz commented on include/wx/qt/glcanvas.h:
```diff
@@ -85,4 +87,27 @@ class WXDLLIMPEXP_GL wxGLCanvas : public wxGLCanvasBase
     wxDECLARE_CLASS(wxGLCanvas);
 };
 
+#include <QtWidgets/QGestureRecognizer>
+#include <QtWidgets/QGestureEvent>
+
+class PanGestureRecognizer : public QGestureRecognizer
```
This doesn't seem to belong here: we really shouldn't have any non-wx private classes in public headers and we try to avoid (and, AFAICS, so far succeed) including platform/toolkit-specific headers from public wx headers.

I think this class could be just moved to `src/qt/glcanvas.cpp`, couldn't it?

transmitterdan replied:
Agreed, no need to put this class in the header file as far as I can tell.  Moving it to glcanvas.cpp now.

vadz commented on include/wx/qt/private/winevent.h:
```diff
@@ -327,6 +330,126 @@ class wxQtEventSignalHandler : public Widget, public wxQtSignalHandler< Handler
 
+    virtual bool event(QEvent *event)
+    {
+        if (event->type() == QEvent::Gesture){
```
It would be great if you could please use wx code style and put braces on their own lines, i.e.
```suggestion
        if (event->type() == QEvent::Gesture)
        {
```
(here and below).

transmitterdan replied:
Agreed. Will fix.

vadz commented on src/common/intl.cpp:
```diff
@@ -1898,7 +1898,8 @@ wxString GetDateFormatFromLangInfo(wxLocaleInfo index)
 wxString wxLocale::GetInfo(wxLocaleInfo index, wxLocaleCategory cat)
 {
 // TODO: as of 2014 Android doesn't has complete locale support (use java api)
-#if !(defined(__WXQT__) && defined(__ANDROID__))
+// DONE: Supported by NDK21d
+//#if !(defined(__WXQT__) && defined(__ANDROID__))
```
Could you please remove these comments completely rather than leaving them if they're not relevant any longer?

transmitterdan replied:
Agreed. Will fix.

vadz commented on src/qt/glcanvas.cpp:
```diff
@@ -378,6 +378,15 @@ wxGLCanvas::wxGLCanvas(wxWindow *parent,
+{
+        // Avoid sending further signals (i.e. if deleting the current page)
+        m_qtWindow->blockSignals(true);
+        // Reset the pointer to avoid handling pending event and signals
+        //QtStoreWindowPointer( GetHandle(), NULL );
+
```
Please don't leave commented out code, it should either be uncommented or removed.

transmitterdan replied:
Agreed. Will fix.

vadz commented on src/qt/glcanvas.cpp:
```diff
@@ -378,6 +378,15 @@ wxGLCanvas::wxGLCanvas(wxWindow *parent,
 
+wxGLCanvas::~wxGLCanvas()
+{
+        // Avoid sending further signals (i.e. if deleting the current page)
+        m_qtWindow->blockSignals(true);
```
Another style comment: indentation seems wrong here, i.e.
```suggestion
    // Avoid sending further signals (i.e. if deleting the current page)
    m_qtWindow->blockSignals(true);
```

transmitterdan replied:
Agreed. Will fix.

vadz commented on src/qt/glcanvas.cpp:
```diff
@@ -425,14 +438,26 @@ bool wxGLCanvas::ConvertWXAttrsToQtGL(const int *wxattrs, QGLFormat &format)
     if (!wxattrs)
         return true;
-    return true;
+    //return true;
```
```suggestion
```

transmitterdan replied:
Agreed. Will fix.

vadz commented on src/qt/glcanvas.cpp:
```diff
@@ -425,14 +438,26 @@ bool wxGLCanvas::ConvertWXAttrsToQtGL(const int *wxattrs, QGLFormat &format)
 
+    // Enable multisampling aka MSAA
+
+    format.setSampleBuffers( true );
+    format.setSamples( 4 );
```
This is normally defined by the attributes used, why should we hard code this here?

vadz replied:
I'm still not sure about this one, could you please check it?

transmitterdan replied:
I am checking.  Will take a little time to find the answers to these two questions.  In mean time I try to clean up all the style and formatting issues.

vadz commented on src/qt/window.cpp:
```diff
@@ -312,8 +337,14 @@ wxWindowQt::~wxWindowQt()
+        QtStoreWindowPointer( GetHandle(), NULL );
 
-    delete m_qtWindow;
+        // Delete QWidget when control return to event loop (safer)
+        m_qtWindow->deleteLater();
```
I'm really not sure about this change. It seems wrong to have to do it, normally we shouldn't get any events after the window is destroyed. Doing this probably means that the sequence of events is not the same when using wxQt and in all the other ports.

Are you sure this is the right solution? Do we have a simple example showing what goes wrong without this commit?

vadz commented on src/qt/window.cpp:
```diff
@@ -1680,3 +1716,31 @@ QPainter *wxWindowQt::QtGetPainter()
+
+bool wxWindowQt::EnableTouchEvents(int eventsMask)
+{
+    if(GetHandle()){
```
This probably should be
```suggestion
    wxCHECK_MSG( GetHandle(), false, "can't be called before creating the window" );
```
instead of silently doing nothing in this case.

transmitterdan replied:
Agreed. Will fix.

vadz reviewed (dismissed):
Thank you for your changes!

Mostly they're good, but the following would need to be done before merging this:

1. Reformat and remove commented out code.
2. Move `PanGestureRecognizer` out of public header (i.e. either a private header or the source file).
3. Check whether hard-coding some GL attributes is really right.
4. Better understand/explain the problem with the "orphan events".

Could you please look at this? TIA!

Commit:
Make class PanGestureRecognizer local to src/qt/window.cpp
Add runtime debug check to wxWindowQt::EnableTouchEvents(int eventsMask)
Clean up formatting and style

Commit:
Remove all trailing whitespace

vadz commented on include/wx/qt/glcanvas.h:
```diff
@@ -85,4 +87,7 @@ class WXDLLIMPEXP_GL wxGLCanvas : public wxGLCanvasBase
     wxDECLARE_CLASS(wxGLCanvas);
 };
 
+#include <QtWidgets/QGestureRecognizer>
+#include <QtWidgets/QGestureEvent>
+
```
Sorry to insist, but are those still required here or could they be moved to the `.cpp` file too? We'd rather not include Qt (or any other toolkit/platform) headers from wx public headers, this can result in unexpected symbol clashes.

transmitterdan replied:
Ok

vadz commented on include/wx/qt/private/winevent.h:
```diff
@@ -327,6 +330,131 @@ class wxQtEventSignalHandler : public Widget, public wxQtSignalHandler< Handler
+        wxWindow *win = wxWindow::QtRetrieveWindowPointer( this );
+
+        if (gesture->state() == Qt::GestureFinished) {
+            if(win){
```
Just another nitpick
```suggestion
            if (win)
            {
```

transmitterdan replied:
Ok

vadz commented on include/wx/qt/private/winevent.h:
```diff
@@ -327,6 +330,131 @@ class wxQtEventSignalHandler : public Widget, public wxQtSignalHandler< Handler
+            event->accept();
+        }
+
+        else {
```
```suggestion
        else
        {
```

transmitterdan replied:
Ok

vadz commented on include/wx/qt/private/winevent.h:
```diff
@@ -327,6 +330,131 @@ class wxQtEventSignalHandler : public Widget, public wxQtSignalHandler< Handler
+    {
+        wxWindow *win = wxWindow::QtRetrieveWindowPointer( this );
+
+        if(win){
```
```suggestion
        if (win)
        {
```

transmitterdan replied:
Ok

vadz commented on include/wx/qt/private/winevent.h:
```diff
@@ -327,6 +330,131 @@ class wxQtEventSignalHandler : public Widget, public wxQtSignalHandler< Handler
+
+            evp.SetDelta( wxQtConvertPoint( delta ) );
+
+            switch(gesture->state()){
```
```suggestion
            switch (gesture->state())
            {
```

transmitterdan replied:
Ok

vadz commented on include/wx/qt/private/winevent.h:
```diff
@@ -327,6 +330,131 @@ class wxQtEventSignalHandler : public Widget, public wxQtSignalHandler< Handler
+    void pinchTriggered(QPinchGesture *gesture, QEvent *event)
+    {
+        wxWindow *win = wxWindow::QtRetrieveWindowPointer( this );
+        if(win){
```
```suggestion
        if (win)
        {
```

transmitterdan replied:
Ok

vadz commented on include/wx/qt/private/winevent.h:
```diff
@@ -327,6 +330,131 @@ class wxQtEventSignalHandler : public Widget, public wxQtSignalHandler< Handler
+            evp.SetPosition( wxQtConvertPoint( center_point ) );
+            evp.SetZoomFactor( this_sf);
+
+            switch(gesture->state()){
```
```suggestion
            switch (gesture->state())
            {
```

transmitterdan replied:
Ok

vadz commented on src/common/intl.cpp:
```diff
@@ -1898,7 +1898,7 @@ wxString GetDateFormatFromLangInfo(wxLocaleInfo index)
 wxString wxLocale::GetInfo(wxLocaleInfo index, wxLocaleCategory cat)
 {
 // TODO: as of 2014 Android doesn't has complete locale support (use java api)
-#if !(defined(__WXQT__) && defined(__ANDROID__))
+// DONE: Supported by NDK21d
```
Let's just remove _both_ of these lines.

transmitterdan replied:
Ok

vadz commented on src/common/intl.cpp:
```diff
@@ -1940,7 +1940,7 @@ wxString wxLocale::GetInfo(wxLocaleInfo index, wxLocaleCategory cat)
             wxFAIL_MSG( "unknown wxLocaleInfo value" );
     }
-#endif
+//#endif
```
And this one
```suggestion
```

transmitterdan replied:
Ok

vadz reviewed (commented):
Thanks for the updates, unfortunately the 2 more important questions (i.e. not counting the trivial style etc issues) still remain unanswered.

Commit:
Clean up more style and formatting

bdbcat committed:
Implement additional GL attribute setting logic

bdbcat committed:
Remove Qt deleteLater() invocation.

bdbcat committed:
Merge remote-tracking branch 'upstream/qtPatches' into qtPatches

bdbcat committed:
Allow for presence of GETHOSTBYADDR without GETHOSTBYNAME

bdbcat committed:
Support wxGLCanvas attributes WX_GL_SAMPLE_BUFFERS, WX_GL_SAMPLES, WX_GL_MAJOR_VERSION

bdbcat committed:
Cleanup #include files.

vadz commented on include/wx/qt/window.h:
```diff
@@ -221,6 +221,7 @@ class WXDLLIMPEXP_CORE wxWindowQt : public wxWindowBase
     // itself.
     virtual QWidget* QtGetParentWidget() const { return GetHandle(); }
 
+    virtual bool EnableTouchEvents(int eventsMask);
```
I think this will trigger clang warnings about inconsistent use of `override`. wxQt should be updated to use it everywhere where it makes sense first, before starting to use it anywhere...

MaartenBent replied:
There are already lots of `wxOVERRIDE` in this header, why would this one be different?

vadz replied:
Oops, yes, you're totally right, sorry. I've only looked at `QtGetParentWidget()` above and recalled my previous problems with adding `wxOVERRIDE` to the (other) wxQt headers which resulted in this problem without realizing that this function wasn't actually overriding anything but was a new virtual in this class.

Sorry again, please disregard the comment above (not deleting it just to avoid confusion).

transmitterdan replied:
Thanks, I will come back to this soon.  We are fixing the hard coded bits first.

bdbcat committed:
Refactor ~wxWindowQt to avoid accessing deleted objects.

bdbcat committed:
Remove redundant include files and unused private member.

bdbcat committed:
Update submodule expat

bdbcat committed:
Update submodule png

bdbcat committed:
Update submodule tiff

bdbcat committed:
Remove trailing white space.

transmitterdan committed:
Merge pull request #4 from bdbcat/qtPatches

Qt patches

transmitterdan committed:
Add wxOVERRIDE and clean up whitespace errors

transmitterdan committed:
Add wxOVERRIDE and remove clang warning

transmitterdan committed:
wxOVERRIDE of SetFont() was causing compiler errors in MinGW

transmitterdan commented:
Thanks to everyone for the code review.  Could everyone (including @bdbcat) do a final review of the diff with master for a final QA check? I want to close this PR and resubmit a clean one that implements the new Qt features.  Again, thank you for helping us get this done.

vadz commented on src/qt/window.cpp:
```diff
@@ -335,6 +361,11 @@ bool wxWindowQt::Create( wxWindowQt * parent, wxWindowID id, const wxPoint & pos
                 QtSetScrollBar( wxHORIZONTAL );
             if ( style & wxVSCROLL )
                 QtSetScrollBar( wxVERTICAL );
+
+            //  Allow processing of QGesture events, if generated upstream.
+            m_qtWindow->setAttribute(Qt::WA_AcceptTouchEvents);
+            m_qtWindow->grabGesture(Qt::PanGesture);
+
```
Do we really need to do it here in addition to doing it in `EnableTouchEvents()`?

bdbcat replied:
Good catch.  Those two lines may be safely deleted.

vadz commented on src/qt/glcanvas.cpp:
```diff
@@ -505,14 +539,16 @@ bool wxGLCanvas::ConvertWXAttrsToQtGL(const int *wxattrs, QGLFormat &format)
 
             case WX_GL_SAMPLE_BUFFERS:
                 format.setSampleBuffers(v);
-                // can we somehow indicate if it's not supported?
```
Not sure why these comments were removed, it doesn't look like the problem they concern was addressed, i.e. we still don't seem to check whether the value is actually supported.

bdbcat replied:
Not sure where these comments came from, maybe original qt port, where some options were not supported, and the case statement was a simple break.  Anyway, the code supports setting the options indicated now.  So the comments no longer mean much.

vadz commented on src/qt/glcanvas.cpp:
```diff
@@ -348,6 +350,30 @@ bool wxGLContext::SetCurrent(const wxGLCanvas&) const
+
+   void reset (QGesture *pGesture);
+
+   uint m_started;
```
This doesn't seem to be actually used anywhere (its use is commented out), could it please be removed completely?

It also looks like it ought to be a `bool` if it were used.

bdbcat replied:
Done.

vadz commented on src/qt/glcanvas.cpp:
```diff
@@ -411,6 +443,9 @@ bool wxGLCanvas::Create(wxWindow *parent,
 
     m_qtWindow = new wxQtGLWidget(parent, this, format);
 
+    QGestureRecognizer* pPanRecognizer = new PanGestureRecognizer();
+    QGestureRecognizer::registerRecognizer(pPanRecognizer);
+
```
I don't know if there is any drawback in doing this unconditionally or if we should only do when touch events are enabled.

Maybe a comment here could help clarify things?

bdbcat replied:
If a custom gesture recognizer is required for a particular type of wxWindow, we need to create and register it.  Theoretically, we could implement EnableTouchEvents(int eventsMask) in this class, which is a wxWindow subclass, and defer creation of the recognizer to that point.  However, Qt provides no interface to determine if a particular recognizer has already been created and registered.  So we might have to resort to some static variable to retain the existence state of the custom recognizer.  Not really a good idea...
A comment in the code could say:
//  Create and register a custom recognizer, available to all instances of this class.

I cannot see any drawback to always creating the recognizer.  If it is never enabled (by EnableTouchEvents()), then it will be dormant and automatically deleted on app exit.

vadz reviewed (commented):
Thanks for the update!

transmitterdan commented:
Closed via https://github.com/wxWidgets/wxWidgets/pull/2168
